### PR TITLE
fix[codegen]: fix non-memory reason strings

### LINKED ISSUE #3787: `Raise` and `Assert` with non-memory msg lead to panicking
### Version Information

* vyper Version (output of `vyper --version`): 4b4e188ba83d28b5dd6ff66479e7448e5b925030
* OS: OSX
* Python Version (output of `python --version`): 3.11.4

### What's your issue about?

`Raise` and `Assert` make the compiler crash if the reason string is not a memory variable.

For example, compiling:

```vyper
a:String[1]
@external
def foo():
    self.a = "a"
    raise self.a
```

leads to `vyper.exceptions.CodegenPanic: unhandled exception 'int' object has no attribute 'typ', parse_Raise`

The issue comes from the fact that `make_byte_array_copier` is given a memory pointer and not an IR node in `Stmt._assert_reason()`

https://github.com/vyperlang/vyper/blob/4b4e188ba83d28b5dd6ff66479e7448e5b925030/vyper/codegen/stmt.py#L180-L183

### How can it be fixed?

```bash
+++ b/vyper/codegen/stmt.py
@@ -179,7 +179,8 @@ class Stmt:
         # TODO maybe use ensure_in_memory
         if msg_ir.location != MEMORY:
             buf = self.context.new_internal_variable(msg_ir.typ)
-            instantiate_msg = make_byte_array_copier(buf, msg_ir)
+            dst = IRnode.from_list(buf, typ=msg_ir.typ, location=MEMORY)
+            instantiate_msg = make_byte_array_copier(dst, msg_ir)
         else:
             buf = _get_last(msg_ir)
             if not isinstance(buf, int):
```

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Fix #3787, fix https://github.com/vyperlang/vyper/issues/3251

### How I did it

Made recommended fix.

### How to verify it

See tests

### Commit message

```
fix code generation for non-memory reason strings. previously, the byte
array copier would fail trying to produce code for the buf, which was an
int rather than an IRnode. this commit cleans up the revert generation
code, allocates a fresh buffer to avoid memory cleanliness issues, and
refactors the code to use abi_encoder code instead of handrolling the
ABI encoding routine.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/codegen/stmt.py [hunk 1]
```diff
@@ -142,7 +142,8 @@ def _get_last(ir):
         # TODO maybe use ensure_in_memory
         if msg_ir.location != MEMORY:
             buf = self.context.new_internal_variable(msg_ir.typ)
-            instantiate_msg = make_byte_array_copier(buf, msg_ir)
+            dst = IRnode.from_list(buf, typ=msg_ir.typ, location=MEMORY)
```
Comment (charles-cooper) [inline vyper/codegen/stmt.py:145]: magic!

Comment (charles-cooper) [reply vyper/codegen/stmt.py:145]: does `ensure_in_memory` work here btw?

Comment (tserg) [reply vyper/codegen/stmt.py:145]: I tried doing this but the same error is thrown:
```
if msg_ir.location != MEMORY:
    buf = self.context.new_internal_variable(msg_ir.typ)
    instantiate_msg = ensure_in_memory(msg_ir, self.context)
else:
    ...
```

Comment (charles-cooper) [reply vyper/codegen/stmt.py:145]: uhm, i think if we use ensure_in_memory we don't need to branch

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**